### ISSUE TITLE
Keypair generated before ce4b5d6 can't be used at master

### ISSUE THREAD

**Issue by ronaldtse** [riboseinc/rnp#335]
## Description

Reported by @catap in #330 . Key pair that was generated before https://github.com/riboseinc/rnp/commit/ce4b5d60378769a3a8b98a46e2098467e7299068 can't be used at master.

> I found first commit that's broke decrypt: ce4b5d6
> If I use a keypair and encrypted file that was created by ce4b5d6, it works at master.
> If I use a keypair and encrypted file that was created by 68b3ef2 (before ce4b5d6), I can't decrypt the file by master.

## Steps to Reproduce

1. Generate key pair before ce4b5d6
2. Encrypt file in master
3. Decryption of file at master fails

( @catap can you confirm these steps?) 

## Expected Behavior

Encryption/decryption of files of any key pair generated before or after ce4b5d6 should work.

## Actual Behavior

> If I use a keypair and encrypted file that was created by 68b3ef2 (before ce4b5d6), I can't decrypt the file by master.

**Comment by ronaldtse** [comment]
@randombit this bug relates to a commit ce4b5d6 that is about Botan key generation, would you be able to help with it? Thanks!

**Comment by randombit** [comment]
Interesting. Yes I'll look at this later today.

**Comment by randombit** [comment]
I believe the problem is in the older keys, and the new behavior is correct. The problem was triggered by this change in ce4b5d6:

```
diff --git a/src/lib/packet-parse.c b/src/lib/packet-parse.c
index f3001d4..09ade66 100644
--- a/src/lib/packet-parse.c
+++ b/src/lib/packet-parse.c
@@ -2483,9 +2483,6 @@ parse_seckey(pgp_content_enum tag, pgp_region_t *region, pgp_stream_t *stream)
                        return 0;
                }
 
-               /* Hardcoded SHA1 for just now */
-               pkt.u.seckey.hash_alg = PGP_HASH_SHA1;
-
                 if(pkt.u.seckey.s2k_specifier == PGP_S2KS_SIMPLE)
```

Basically, in `rnp` before ce4b5d6, no matter what hash was indicated in the private key packet, SHA1 was used for the S2K derivation. This is obviously incorrect if the hash stored was not SHA1. But, the default hash set by `rnpkeys` is SHA-256. So the result is, keys generated by rnpkeys before that revision save the key using SHA-1 for the S2K, but indicate that they used SHA-256! So I expect they would not be decryptable by `gpg` either, assuming there was some way to export private keys.

I double checked this by (with an older version) using `./src/rnpkeys/rnpkeys --hash SHA1 --generate-key` - and confirmed this key could successfully be decrypted with master, because the hash used for S2K derivation and the hash indicated in the key format now match.

So tl;dr I think the current behavior is correct and the old keys are in fact badly formed. If there is a need to support reading older keys we can probably do this by falling back to SHA-1 S2K if the indicated hash is not SHA-1 and the decrypt failed. Or alternately (and maybe better) add an option to `rnpkeys` to migrate the keys - try to decrypt using the indicated hash, if it fails, try SHA-1, if that works patch up the private key to store the correct hash in the S2K field. That way if/when private key export is available, we don't have to worry about exporting invalid keys that won't be understood by gpg.

**Comment by catap** [comment]
@randombit anyway, I found this problem as side effect then I tried to realise why same cycle (encode file and decode it) doesn't work by `key-public.asc` and `key-private.asc` from test folder that was a part of original netgpg.

**Comment by catap** [comment]
@ronaldtse it is a question about old user. So, if we haven't got any users yet, we can ignore it.

But if we have users who uses an old release we might broke rnp for them with new release.

**Comment by ronaldtse** [comment]
There are a total of 3 people I don't know here who aren't in development of rnp. https://github.com/riboseinc/rnp/stargazers

Maybe we can wait for someone to complain before fixing this. I imagine whoever is using rnp would probably be using GnuPG keys...

**Comment by randombit** [comment]
@ronaldtse The new (correct) behavior has been in place over a year with no issue raised, so I think no production keys were affected. I believe this issue can be closed, there is no need to implement any kind of workaround.

**Comment by ronaldtse** [comment]
Thanks @randombit for the confirmation!
